File: ios/SsdidWallet/Domain/Notify/NotifyManager.swift
import Foundation
import CryptoKit
import UserNotifications

/// Orchestrates SSDID Notify service integration: inbox registration, APNs device token
/// updates, per-identity mailbox lifecycle, and pending notification retrieval.
///
/// Inbox credentials (inbox_id / inbox_secret) are stored in the Keychain.
/// Non-sensitive metadata (inbox_id) is also mirrored in UserDefaults for quick reads.
@MainActor
final class NotifyManager {

    // MARK: - Keychain aliases

    private static let inboxSecretAlias = "notify.inbox.secret"
    private static let inboxIdAlias = "notify.inbox.id"
    private static let inboxIdDefaultsKey = "notify.inbox.id"

    // MARK: - Dependencies

    private let api: NotifyApi
    private let keychainManager: KeychainManager

    // MARK: - State

    /// Guards against concurrent calls to ensureInboxRegistered causing a double-registration.
    /// Safe without a lock because this class is @MainActor-isolated.
    private var isRegistering = false

    // MARK: - Identity Resolution

    /// Closure that returns the current list of identities.
    /// Set by the app layer (ServiceContainer or SsdidWalletApp) to enable demux.
    var identityProvider: (() async -> [Identity])?

    // MARK: - Init

    init(api: NotifyApi, keychainManager: KeychainManager) {
        self.api = api
        self.keychainManager = keychainManager
    }

    // MARK: - Inbox

    /// Ensures the device has a registered inbox with the Notify service.
    ///
    /// On first launch this registers a new inbox (without a device token — APNs
    /// registration happens separately). On subsequent launches it returns immediately
    /// because the credentials are already stored in the Keychain.
    func ensureInboxRegistered() async throws {
        guard !isRegistering else { return }
        isRegistering = true
        defer { isRegistering = false }
        guard (try? keychainManager.loadSecret(alias: Self.inboxSecretAlias)) == nil else { return }

        // Register with no devices; APNs token is supplied later via registerAPNsToken(_:).
        let response = try await api.registerInbox(
            request: InboxRegisterRequest(devices: [])
        )

        try keychainManager.saveSecret(alias: Self.inboxSecretAlias, value: response.inboxSecret)
        try keychainManager.saveSecret(alias: Self.inboxIdAlias, value: response.inboxId)
        UserDefaults.standard.set(response.inboxId, forKey: Self.inboxIdDefaultsKey)
    }

    /// Converts the raw APNs device token to a hex string and registers (or updates)
    /// it with the Notify service. Safe to call every time the app receives a new token.
    func registerAPNsToken(_ tokenData: Data) async throws {
        guard let inboxSecret = try keychainManager.loadSecret(alias: Self.inboxSecretAlias) else {
            // Inbox not yet registered; token will be registered after ensureInboxRegistered.
            return
        }

        let hexToken = tokenData.map { String(format: "%02x", $0) }.joined()
        let device = NotifyDevice(platform: "ios", token: hexToken)

        try await api.updateDevices(
            inboxSecret: inboxSecret,
            request: InboxUpdateDevicesRequest(devices: [device])
        )
    }

    // MARK: - Mailboxes

    /// Creates a mailbox for the given identity.
    ///
    /// The mailbox_id is deterministic: `"mbx_"` followed by the first 16 characters of
    /// the URL-safe base64 (no padding) encoding of SHA-256(did), providing
    /// correlation resistance while remaining stable across reinstalls.
    func createMailbox(for identity: Identity) async throws {
        guard let inboxSecret = try keychainManager.loadSecret(alias: Self.inboxSecretAlias) else {
            throw NotifyError.inboxNotRegistered
        }

        let inboxId = try loadInboxId()

        let mailboxId = mailboxIdForDid(identity.did)

        try await api.createMailbox(
            inboxSecret: inboxSecret,
            request: MailboxCreateRequest(inboxId: inboxId, mailboxId: mailboxId)
        )
    }

    /// Deletes the mailbox associated with the given identity.
    /// Call this when an identity is deactivated or removed.
    func deleteMailbox(for identity: Identity) async throws {
        guard let inboxSecret = try keychainManager.loadSecret(alias: Self.inboxSecretAlias) else {
            throw NotifyError.inboxNotRegistered
        }

        let mailboxId = mailboxIdForDid(identity.did)

        try await api.deleteMailbox(inboxSecret: inboxSecret, mailboxId: mailboxId)
    }

    // MARK: - Pending Notifications

    /// Fetches all pending notifications from the Notify service.
    /// Call this on reconnect or when the app returns to foreground.
    func fetchPending() async throws -> [NotifyNotification] {
        guard let inboxSecret = try keychainManager.loadSecret(alias: Self.inboxSecretAlias) else {
            throw NotifyError.inboxNotRegistered
        }

        let response = try await api.fetchPending(inboxSecret: inboxSecret)
        return response.notifications
    }

    /// Acknowledges a notification by deleting it from the Notify service.
    /// Call this after the notification has been fully processed.
    func ackPending(notificationId: String) async throws {
        guard let inboxSecret = try keychainManager.loadSecret(alias: Self.inboxSecretAlias) else {
            throw NotifyError.inboxNotRegistered
        }

        try await api.ackNotification(inboxSecret: inboxSecret, notificationId: notificationId)
    }

    // MARK: - Fetch & Demux

    /// Fetches all pending notifications, resolves each to its owning identity,
    /// posts a local notification with identity context, and acknowledges.
    func fetchAndDemux() async throws {
        let pending = try await fetchPending()
        guard !pending.isEmpty else { return }

        // Build reverse map: mailboxId -> identity name
        let identities = await identityProvider?() ?? []
        var mailboxToName: [String: String] = [:]
        for identity in identities {
            let mbxId = mailboxIdForDid(identity.did)
            mailboxToName[mbxId] = identity.name
        }

        let center = UNUserNotificationCenter.current()

        for notification in pending {
            let identityName = mailboxToName[notification.mailboxId]

            let content = UNMutableNotificationContent()
            content.title = identityName.map { "SSDID — \($0)" } ?? "SSDID"
            content.body = notification.payload
            content.sound = .default

            let request = UNNotificationRequest(
                identifier: notification.notificationId,
                content: content,
                trigger: nil // Deliver immediately
            )
            try? await center.add(request)

            // Acknowledge — failure means re-fetch next time, acceptable
            try? await ackPending(notificationId: notification.notificationId)
        }
    }

    // MARK: - Private Helpers

    /// Reads the inbox ID with fallback: UserDefaults first, then Keychain.
    private func loadInboxId() throws -> String {
        if let id = UserDefaults.standard.string(forKey: Self.inboxIdDefaultsKey) {
            return id
        }
        if let id = try keychainManager.loadSecret(alias: Self.inboxIdAlias) {
            // Repair UserDefaults from Keychain (e.g. after a restore).
            UserDefaults.standard.set(id, forKey: Self.inboxIdDefaultsKey)
            return id
        }
        throw NotifyError.inboxNotRegistered
    }

    /// Derives a stable mailbox ID from a DID string.
    /// Format: "mbx_" + base64url-nopad(SHA-256(did))[0..<16]
    private func mailboxIdForDid(_ did: String) -> String {
        let digest = SHA256.hash(data: Data(did.utf8))
        let hashData = Data(digest)
        let encoded = hashData.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
        let prefix = String(encoded.prefix(16))
        return "mbx_\(prefix)"
    }
}

// MARK: - NotifyError

enum NotifyError: Error, LocalizedError {
    case inboxNotRegistered

    var errorDescription: String? {
        switch self {
        case .inboxNotRegistered:
            return "Notify inbox has not been registered yet. Call ensureInboxRegistered() first."
        }
    }
}

File: android/app/src/main/java/my/ssdid/wallet/domain/notify/NotifyManager.kt
package my.ssdid.wallet.domain.notify

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import my.ssdid.wallet.domain.logging.NoOpLogger
import my.ssdid.wallet.domain.logging.SsdidLogger
import my.ssdid.wallet.domain.model.Identity
import my.ssdid.wallet.domain.transport.NotifyApi
import my.ssdid.wallet.domain.transport.dto.CreateMailboxRequest
import my.ssdid.wallet.domain.transport.dto.NotifyDevice
import my.ssdid.wallet.domain.transport.dto.PendingNotification
import my.ssdid.wallet.domain.notify.LocalNotificationStore
import my.ssdid.wallet.domain.transport.dto.RegisterInboxRequest
import my.ssdid.wallet.domain.transport.dto.UpdateDevicesRequest
import java.security.MessageDigest
import java.util.Base64
import kotlin.jvm.Volatile

/**
 * Orchestrates all interactions with the SSDID Notify service.
 *
 * Responsibilities:
 * - Register an inbox on first launch and persist credentials via [NotifyStorage].
 * - Create/delete per-identity mailboxes (correlation-free notifications).
 * - Fetch and acknowledge pending notifications after a reconnect.
 *
 * The mailbox_id is deterministic: "mbx_" + first 16 chars of Base64url(SHA-256(did.utf8)).
 * This allows recreating it without stored state.
 */
class NotifyManager(
    private val notifyApi: NotifyApi,
    private val storage: NotifyStorage,
    private val dispatcher: NotifyDispatcher,
    private val localNotificationStorage: LocalNotificationStore,
    private val logger: SsdidLogger = NoOpLogger()
) {

    private val registrationMutex = Mutex()

    @Volatile
    private var knownIdentities: List<Identity> = emptyList()

    /**
     * Updates the list of known identities used for mailbox -> identity resolution.
     * Call this whenever identities change (create, delete, foreground refresh).
     */
    fun updateKnownIdentities(identities: List<Identity>) {
        knownIdentities = identities
    }

    /**
     * Fetches pending notifications, resolves each to its owning identity via
     * mailbox_id reverse lookup, dispatches via [dispatcher], and acknowledges.
     */
    suspend fun fetchAndDemux() {
        val pending = fetchPending()
        if (pending.isEmpty()) return

        // Build reverse map: mailbox_id -> identity name
        val mailboxToName = knownIdentities.associate { identity ->
            mailboxIdFor(identity) to identity.name
        }

        for (notification in pending) {
            val identityName = mailboxToName[notification.mailboxId]

            // Save locally first
            localNotificationStorage.save(
                LocalNotification(
                    id = notification.notificationId,
                    mailboxId = notification.mailboxId,
                    identityName = identityName,
                    payload = notification.payload,
                    priority = notification.priority,
                    receivedAt = notification.receivedAt ?: "",
                    isRead = false
                )
            )

            try {
                dispatcher.dispatch(identityName, notification)
            } catch (_: Exception) {
                // Dispatch failure is non-fatal
            }
            try {
                ackPending(notification.notificationId)
            } catch (_: Exception) {
                // Ack failure means it'll be re-fetched next time — acceptable
            }
        }
    }

    /**
     * Ensures an inbox exists for this device. Idempotent — safe to call on every launch.
     * If no inbox is registered yet, creates one with no device tokens attached (tokens are
     * added separately once a FCM/UnifiedPush token is available).
     *
     * A Mutex prevents TOCTOU races when called concurrently (e.g. FCM token arrives at
     * the same time as the app-startup call).
     *
     * @return the inbox_id, or null if registration failed.
     */
    suspend fun ensureInboxRegistered(): String? = registrationMutex.withLock {
        val existing = storage.getInboxId()
        if (existing != null) return@withLock existing

        try {
            // Register with an empty devices list; tokens are updated via updateDeviceToken.
            val response = notifyApi.registerInbox(RegisterInboxRequest(devices = emptyList()))
            storage.saveInboxCredentials(response.inboxId, response.inboxSecret)
            logger.info(TAG, "Inbox registered: ${response.inboxId}")
            response.inboxId
        } catch (e: Exception) {
            logger.error(TAG, "Failed to register inbox", e)
            null
        }
    }

    /**
     * Updates the push device token on an existing inbox.
     * Call this whenever a new FCM/UnifiedPush token is received.
     *
     * @param platform e.g. "android"
     * @param token the push token (e.g. FCM token or UnifiedPush endpoint URL)
     */
    suspend fun updateDeviceToken(platform: String, token: String) {
        val (inboxId, bearer) = credentials() ?: return
        try {
            notifyApi.updateDevices(
                bearerSecret = bearer,
                request = UpdateDevicesRequest(
                    devices = listOf(NotifyDevice(platform = platform, token = token))
                )
            )
            logger.info(TAG, "Device token updated for inbox $inboxId")
        } catch (e: Exception) {
            logger.error(TAG, "Failed to update device token", e)
        }
    }

    /**
     * Creates a mailbox for the given identity. The mailbox_id is derived deterministically
     * from the identity's DID so it can always be recreated.
     *
     * Safe to call multiple times — the server is expected to be idempotent on duplicate
     * mailbox_id for the same inbox.
     */
    suspend fun createMailbox(identity: Identity) {
        val (inboxId, bearer) = credentials() ?: return
        val mailboxId = mailboxIdFor(identity)
        try {
            notifyApi.createMailbox(
                bearerSecret = bearer,
                request = CreateMailboxRequest(inboxId = inboxId, mailboxId = mailboxId)
            )
            logger.info(TAG, "Mailbox created: $mailboxId for ${identity.did}")
        } catch (e: Exception) {
            logger.error(TAG, "Failed to create mailbox for ${identity.did}", e)
        }
    }

    /**
     * Deletes the mailbox associated with the given identity (e.g. when it is deactivated).
     */
    suspend fun deleteMailbox(identity: Identity) {
        val (_, bearer) = credentials() ?: return
        val mailboxId = mailboxIdFor(identity)
        try {
            notifyApi.deleteMailbox(bearerSecret = bearer, mailboxId = mailboxId)
            logger.info(TAG, "Mailbox deleted: $mailboxId for ${identity.did}")
        } catch (e: Exception) {
            logger.error(TAG, "Failed to delete mailbox for ${identity.did}", e)
        }
    }

    /**
     * Fetches all pending notifications from the server.
     * Call this on reconnect / app foreground to drain queued messages.
     *
     * @return list of pending notifications, or empty list on failure.
     */
    suspend fun fetchPending(): List<PendingNotification> {
        val (_, bearer) = credentials() ?: return emptyList()
        return try {
            notifyApi.fetchPending(bearerSecret = bearer).notifications
        } catch (e: Exception) {
            logger.error(TAG, "Failed to fetch pending notifications", e)
            emptyList()
        }
    }

    /**
     * Acknowledges a notification after it has been processed, removing it from the server queue.
     */
    suspend fun ackPending(notificationId: String) {
        val (_, bearer) = credentials() ?: return
        try {
            notifyApi.ackPending(bearerSecret = bearer, notificationId = notificationId)
        } catch (e: Exception) {
            logger.error(TAG, "Failed to ack notification $notificationId", e)
        }
    }

    // ---------- Helpers ----------

    /**
     * Returns inbox_id and a "Bearer {inbox_secret}" header value, or null if not registered.
     */
    private suspend fun credentials(): Pair<String, String>? {
        val inboxId = storage.getInboxId() ?: return null
        val secret = storage.getInboxSecret() ?: return null
        return inboxId to "Bearer $secret"
    }

    /**
     * Derives a stable mailbox_id from an identity's DID.
     * Format: "mbx_" + first 16 chars of Base64url-nopad(SHA-256(did.utf8))
     * (yields ~96 bits of entropy, sufficient for correlation resistance).
     */
    private companion object {
        private const val TAG = "NotifyManager"
    }

    internal fun mailboxIdFor(identity: Identity): String {
        val digest = MessageDigest.getInstance("SHA-256")
            .digest(identity.did.toByteArray(Charsets.UTF_8))
        val encoded = Base64.getUrlEncoder().withoutPadding().encodeToString(digest)
        return "mbx_${encoded.take(16)}"
    }

}
